src/nbl/video/CVKLogicalDevice.h

```diff
@@ -124,51 +327,472 @@ class CVKLogicalDevice final : public ILogicalDevice
+        // vkGetBufferMemoryRequirements or equivalent
+        core::smart_refctd_ptr<video::IDriverMemoryAllocation> bufferMemory =
+            allocateDeviceLocalMemory(gpuBuffer->getMemoryReqs());
+
+        if (!bufferMemory)
+            return nullptr;
+
+        ILogicalDevice::SBindBufferMemoryInfo bindBufferInfo = {};
+        bindBufferInfo.buffer = gpuBuffer.get();
+        bindBufferInfo.memory = bufferMemory.get();
+        bindBufferInfo.offset = 0ull;
+        bindBufferMemory(1u, &bindBufferInfo);
+
+        return gpuBuffer;
+    }
+        
     core::smart_refctd_ptr<IGPUShader> createGPUShader(core::smart_refctd_ptr<asset::ICPUShader>&& cpushader) override
     {
-        return nullptr;
+        const asset::ICPUBuffer* source = cpushader->getSPVorGLSL();
+        core::smart_refctd_ptr<asset::ICPUBuffer> clone =
+            core::smart_refctd_ptr_static_cast<asset::ICPUBuffer>(source->clone(1u));
+        if (cpushader->containsGLSL())
+        {
+            return core::make_smart_refctd_ptr<CVulkanShader>(
+                core::smart_refctd_ptr<CVKLogicalDevice>(this), std::move(clone),
+                IGPUShader::buffer_contains_glsl);
+        }
+        else
+        {
+            return core::make_smart_refctd_ptr<CVulkanShader>(core::smart_refctd_ptr<CVKLogicalDevice>(this),
+                std::move(clone));
+        }
+    }
+
+    core::smart_refctd_ptr<IGPUImage> createGPUImage(asset::IImage::SCreationParams&& params) override
+    {
+        VkImageCreateInfo vk_createInfo = { VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO };
+        vk_createInfo.pNext = nullptr; // there are a lot of extensions
+        vk_createInfo.flags = static_cast<VkImageCreateFlags>(params.flags);
+        vk_createInfo.imageType = static_cast<VkImageType>(params.type);
+        vk_createInfo.format = getVkFormatFromFormat(params.format);
+        vk_createInfo.extent = { params.extent.width, params.extent.height, params.extent.depth };
+        vk_createInfo.mipLevels = params.mipLevels;
+        vk_createInfo.arrayLayers = params.arrayLayers;
+        vk_createInfo.samples = static_cast<VkSampleCountFlagBits>(params.samples);
+        vk_createInfo.tiling = static_cast<VkImageTiling>(params.tiling);
+        vk_createInfo.usage = static_cast<VkImageUsageFlags>(params.usage);
+        vk_createInfo.sharingMode = static_cast<VkSharingMode>(params.sharingMode);
+        vk_createInfo.queueFamilyIndexCount = params.queueFamilyIndices->size();
+        vk_createInfo.pQueueFamilyIndices = params.queueFamilyIndices->data();
+        vk_createInfo.initialLayout = static_cast<VkImageLayout>(params.initialLayout);
+
+        VkImage vk_image;
+        if (vkCreateImage(m_vkdev, &vk_createInfo, nullptr, &vk_image) == VK_SUCCESS)
+        {
+            return core::make_smart_refctd_ptr<CVulkanImage>(core::smart_refctd_ptr<CVKLogicalDevice>(this),
+                std::move(params), vk_image);
```

Comment:
you want to query and cache the device memory requirements, just like the IGPUBuffer does